### PR TITLE
feat: Add webhook parser for Sendinblue

### LINKED ISSUE #1463: [NV-1009] - Create webhook parser for Sendinblue (email)
**Why**

To be able to give better answers why a message was not sent to a subscriber we need to get some info from the provider with a webhook. (deliverability status, rejection status etc).

**How**

To do this we need to implement these functions inside of the provider:

```
  getMessageId?: (body: any | any[]) => string[];

  parseEventBody?: (
    body: any | any[],
    identifier: string
  ) => IEmailEventBody | undefined;
```

The `getMessageId` needs to return all identifiers that should be the same value as id that is returned from `sendMessage` with help of this we can connect a webhook to the specific message.

`parseEventBody` will get the full webhook body and the current identifier that the body should be parsed for. If there is nothing to parse the method should return `undefined`. 

All returns that is not `undefined` should be an object that is in this format:

```
status: string;

 date: string;

externalId?: string;

attempts?: number;

response?: string;

// Contains the raw content for message from the provider webhook

row?: string;
```

Note that statuses for email needs to be value from `EmailEventStatusEnum` and sms should have a status from `SmsEventStatusEnum.` 

 Example of how to do it can be found in `SendgridEmailProvider`.

### PR DESCRIPTION
**Pull request by ShaneMaglangit**
**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

Implemented `getMessageId` and `parseEventBody` to parse Sendinblue webhook.

**Why was this change needed?** (You can also link to an open issue here)

Fixes #1463 

**Other information**:

JSON Response [(Reference)](https://developers.sendinblue.com/docs/how-to-use-webhooks) -- Structure reflects real response (tested by triggering real webhook through the platform)
```json
{
  "event": "delivered",
  "email": "example@example.com",
  "id": 26224,
  "date": "YYYY-MM-DD HH:mm:ss",
  "ts": 1598634509,
  "message-id": "<xxxxxxxxxxxx.xxxxxxxxx@domain.com>",
  "ts_event": 1598034509,
  "subject": "Subject Line",
  "tag": "[\"transactionalTag\"]",
  "sending_ip": "185.41.28.109",
  "ts_epoch": 1598634509223,
  "tags": [
    "myFirstTransactional"
  ]
}
```

Events [(Reference)](https://developers.sendinblue.com/reference/createwebhook)
image

### CODE REVIEW

File: providers/sendinblue/src/lib/sendinblue.provider.ts [hunk 1]
```diff
@@ -54,6 +56,75 @@ export class SendinblueEmailProvider implements IEmailProvider {
     };
   }
 
+  getMessageId(body: any | any[]): string[] {
+    if (Array.isArray(body)) {
+      return body.map((item) => item['message-id']);
+    }
+
+    return [body['message-id']];
```
Comment (davidsoderberg) [inline providers/sendinblue/src/lib/sendinblue.provider.ts:64]: this is the same as messageId that is returned from sendMessage?

File: providers/sendinblue/src/lib/sendinblue.provider.ts [hunk 2]
```diff
@@ -54,6 +56,75 @@ export class SendinblueEmailProvider implements IEmailProvider {
...
+      status: status,
+      date: body.date,
+      externalId: body.id,
+      row: body,
+    };
+  }
+
+  private getStatus(event: string): EmailEventStatusEnum | undefined {
+    switch (event) {
+      case 'sent':
```
Comment (davidsoderberg) [inline providers/sendinblue/src/lib/sendinblue.provider.ts:95]: Maybe these can be an internal enum inside of this file to make easy to change in the future?

File: providers/sendinblue/src/lib/sendinblue.provider.ts [hunk 3]
```diff
@@ -54,6 +56,75 @@ export class SendinblueEmailProvider implements IEmailProvider {
...
+
+    const status = this.getStatus(body.event);
+
+    if (status === undefined) {
+      return undefined;
+    }
+
+    return {
+      status: status,
+      date: body.date,
```
Comment (davidsoderberg) [inline providers/sendinblue/src/lib/sendinblue.provider.ts:87]: What format do this date have?

**Comment by davidsoderberg** [comment]
Hi @ShaneMaglangit looks good, would be great if we could test this in the provider spec as well 😄

File: providers/sendinblue/src/lib/sendinblue.provider.ts (hunk 3 above)
Comment (ShaneMaglangit) [reply providers/sendinblue/src/lib/sendinblue.provider.ts:87]: The date was formatted as `YYYY-MM-DD HH:mm:ss`. Is there are standard format that is used for in this? ISO string perhaps?

File: providers/sendinblue/src/lib/sendinblue.provider.ts (hunk 2 above)
Comment (ShaneMaglangit) [reply providers/sendinblue/src/lib/sendinblue.provider.ts:95]: I believe that is contradictory to what we have [discussed recently](https://discord.com/channels/895029566685462578/937309213036785744/1029245883499429899).

However, I do agree and much prefer creating a new `SendinblueEventStatusEnum`.

File: providers/sendinblue/src/lib/sendinblue.provider.ts (hunk 1 above)
Comment (ShaneMaglangit) [reply providers/sendinblue/src/lib/sendinblue.provider.ts:64]: Yes, I confirmed it by manually testing the hook. Here is the screenshot from Beeceptor where the response was sent.

image

**Comment by ShaneMaglangit** [comment]
@davidsoderberg Hey, I pushed some changes on the PR. Would love to hear your thoughts about it.

I modified the event parsing function based to serve more as a ["dedicated mapping function from provider to novu events"](https://discord.com/channels/895029566685462578/937309213036785744/1029280039864373248), or at least the way that I understood it.

I've also added unit tests for the added functions.

**Review by scopsy — APPROVED** [review]
thank you @ShaneMaglangit!

**Pull request merged**